### PR TITLE
Implement PaginatedProposalVotes API

### PR DESCRIPTION
**Pull request by hochiw**
- Added Validator models and resolvers
- Added Validator to ProposalVote
- Handle paginated proposal vote query

refs oursky/likedao#129 (1 of 2)

### CODE REVIEW

**Comment by chihimng** [comment]
as discussed offline, please show validators entries even if user has not delegated to them

**Comment by hochiw** [comment]
please also see the resolver so I can change them altogether

**Comment by hochiw** [comment]
@chihimng updated

File: graphql-schema/proposal.graphql [hunk 1]
```diff
@@ -69,6 +69,19 @@ type ProposalTallyResult {
   outstandingOption: ProposalVoteOption
 }
 
+# Created to handle union proposal voter type
+type StringObject {
```
Comment (chihimng) [inline graphql-schema/proposal.graphql:73]: move to standalone file?

File: graphql-schema/proposal.graphql [hunk 2]
```diff
@@ -92,7 +103,17 @@ input QueryProposalsInput {
   searchTerm: String
 }
 
+input QueryProposalVotesInput {
+  proposalId: ID!
+  # Limit
+  first: Int!
+  # Offset
+  after: Int!
+  delegatedValidators: [String!]
```
Comment (chihimng) [inline graphql-schema/proposal.graphql:112]: `pinnedValidators`? can ignore

**Review by elise-ng — CHANGES_REQUESTED** [review]

**Comment by hochiw** [comment]
Second thought I think it would be better to put the query inside the Proposal object so that we don't have to pass the id again, updating this pr

File: graphql-server/pkg/dataloaders/validator.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+import (
+	godataloader "github.com/cychiuae/go-dataloader"
+	"github.com/oursky/likedao/pkg/models"
+	"github.com/oursky/likedao/pkg/queries"
+)
+
+type ValidatorDataloader interface {
+	LoadValidatorWithInfoByConsensusAddress(address string) (*models.Validator, error)
+	LoadValidatorWithInfoByDelegationAddress(address string) (*models.Validator, error)
```
Comment (chihimng) [inline graphql-server/pkg/dataloaders/validator.go:11]: better name this `BySelfDelegateAddress`?

File: graphql-schema/validator.graphql [hunk 4]
```diff
@@ -0,0 +1,12 @@
+type Validator implements Node {
+  id: ID!
+  consensusAddress: String!
+  operatorAddress: String
+  delegationAddress: String
```
Comment (chihimng) [inline graphql-schema/validator.graphql:5]: should be selfDelegationAddress?

File: graphql-schema/base.graphql [hunk 5]
```diff
@@ -13,3 +13,8 @@ scalar BigFloat
 type StringObject {
   value: String!
 }
+
+enum Sort {
+  asc
```
Comment (chihimng) [inline graphql-schema/base.graphql:18]: enum values should be all caps?

File: graphql-server/pkg/resolvers/proposal.go [hunk 6]
```diff
@@ -87,6 +87,100 @@ func (r *proposalResolver) MyReaction(ctx context.Context, obj *models.Proposal)
...
+	if err != nil {
+		return nil, servererrors.QueryError.NewError(ctx, fmt.Sprintf("failed to load validators: %v", err))
+	}
+
+	// Add validators to result
+	validatorDelegationAddresses := make([]string, 0)
+	for i := range validators.Items {
+		validator := validators.Items[i]
+		if validator.Info != nil && len(validator.Info.ProposalVotes) != 0 {
+			result = append(result, *validator.Info.ProposalVotes[0])
```
Comment (chihimng) [inline graphql-server/pkg/resolvers/proposal.go:111]: why `[0]`? if a validator voted multiple times will the query return latest vote?

File: graphql-server/pkg/resolvers/proposal.go [hunk 7]
```diff
@@ -87,6 +87,100 @@ func (r *proposalResolver) MyReaction(ctx context.Context, obj *models.Proposal)
 	return &reaction.Reaction, nil
 }
 
+func (r *proposalResolver) Votes(ctx context.Context, obj *models.Proposal, input models.QueryProposalVotesInput) (*models.Connection[models.ProposalVote], error) {
```
Comment (chihimng) [inline graphql-server/pkg/resolvers/proposal.go:90]: This logic is quite complicated, best if we add test cases later

**Review by elise-ng — CHANGES_REQUESTED** [review]

File: graphql-server/pkg/resolvers/proposal.go (hunk 6 above)
Comment (hochiw) [reply graphql-server/pkg/resolvers/proposal.go:111]: the vote gets overwritten when voted multiple times
https://github.com/forbole/bdjuno/blob/50e97941f2ae69dfba6e2382c6c16cd641464fd4/database/gov.go#L270-L275

`ProposalVotes` itself has a many-to-one relationship with validators but when the proposal id scope is applied it becomes one-to-one so `len(validator.Info.ProposalVotes)` would always be either 1 or 0 here

maybe think think if there are better ways to handle this

Comment (hochiw) [reply graphql-server/pkg/resolvers/proposal.go:112]: Maybe I change the proposal id scope to singular first to ensure there's only 1 vote

**Comment by hochiw** [comment]
@chihimng updated 🙏

**Comment by hochiw** [comment]
merge first

**Pull request merged**